### PR TITLE
RideShare - Angela - Octos

### PR DESCRIPTION
**Pull request by knockknockhusthere**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I created a lot of helper methods in the driver class for calculations. At first, I calculated total drive time, and total money in seperate methods, and then created a separate one to calculate average per hour. Additionally, after we implemented request_ride in the trip dispatcher class, I created a method that returns a list of the driver's complete trips only.
Describe any examples of composition that you encountered in this project, if any | There were many examples of composition between the classes. For example, every instance of trip has an instance of passenger and driver. Every driver has a list of their trips. Trip Dispatcher has a list of all trips, all drivers, and all passengers. 
Describe a _nominal test_ that you wrote for this assignment. | For the average_rating method in driver class, I tested that the method returns a float between 1.0 and 5.0. 
Describe an _edge case test_ that you wrote for this assignment. | For the average_rating method in driver method, I also tested that in the case of zero trips, the rating would return nil. 
Describe a concept that you gained more clarity on as you worked on this assignment. | I got a better grasp on how classes interact and how initialize works. I didn't know you could create instance variables in the constructor by calling other methods.

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -22,26 +23,55 @@ def initialize(input)
...
         raise ArgumentError.new("Can only add trip instance to trip collection")
       end
-
       @trips << trip
     end
+
+    def ignores_incomplete_rides()
+      complete_rides = @trips.select { |trip| trip.end_time != nil }
+      return complete_rides
+    end
```
Comment (droberts-sea) [inline lib/driver.rb:46]: I really like the idea of this helper method - good job identifying it as a responsibility of `Driver`. I'm not quite sure about the name though - something like `completed_trips` might be a little more concise.

File: lib/passenger.rb [hunk 2]
```diff
@@ -20,5 +20,20 @@ def get_drivers
     def add_trip(trip)
       @trips << trip
     end
-  end
-end
+
+    def total_trips_cost()
+      total_cost = @trips.inject(0) do |total, trip|
+        total + trip.cost
+      end
```
Comment (droberts-sea) [inline lib/passenger.rb:27]: Attempting to run this method on a passenger with an unfinished trip yields an error!

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | yes
Answer comprehension questions | yes
**Wave 1** |
`Trip` has a helper method to calculate duration | yes
`Passenger` has a method to calculate total cost of all trips | yes
`Driver` has a method to calculate total revenue | yes
Complete tests | many interesting cases missing - see inline
**Wave 2** |
`TripDispatcher` has a new method to create trips | yes
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | yes
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | several of your `Passenger` methods fail!
Complete tests | mostly - see inline
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | N/A
Appropriate use of Ruby's `Date` | N/A
Appropriate helper methods were made to help with complex logic | N/A
Complete tests | N/A
**Overall** | Good job overall! Your code is clear and well-written, and you've demonstrated a strong understanding of functional decomposition. I am seeing some big holes in your test coverage.  I've tried to outline these with inline comments, but the general pattern is you get the nominal case, but miss interesting edge or error cases. Make sure you're thinking not just about how your method will succeed, but also about how it might fail.<br><br>I'm excited to see what you come up with for Hotel. Keep up the hard work!

File: specs/driver_spec.rb [hunk 3]
```diff
@@ -72,9 +72,102 @@
...
+    it "returns nil if no trips" do
       driver = RideShare::Driver.new(id: 54, name: "Rogers Bartell IV", vin: "1C9EVBRM0YBC564DZ")
-      driver.average_rating.must_equal 0
+      driver.average_rating.must_equal nil
+    end
+  end
+
+  describe 'calculate_total_rev method' do
+    it 'can return total money driver made' do
+      start_time = Time.parse('2015-05-20T12:14:00+00:00')
```
Comment (droberts-sea) [inline specs/driver_spec.rb:83]: Two interesting cases that apply to all of these describe blocks:

- What if the driver has driven 0 trips?
- What if the driver has an incomplete trip?

File: specs/driver_spec.rb [hunk 4]
```diff
@@ -72,9 +72,102 @@
...
+
+  describe 'total_drive_time' do
+    it 'can return total time driver drove' do
+      start_time = Time.parse('2015-05-20T12:14:00+00:00')
+      end_time = start_time + 25 * 60 # 25 minutes
+      driver = RideShare::Driver.new(id: 54, name: "Rogers Bartell IV", vin: "1C9EVBRM0YBC564DZ")
+      trip1 = RideShare::Trip.new({id: 8, start_time: start_time, end_time: end_time, driver: driver, cost: 15.90, passenger: nil, date: "2016-08-08", rating: 5})
+      trip2 = RideShare::Trip.new({id: 9, start_time: start_time, end_time: end_time, driver: driver, cost: 12.10, passenger: nil, date: "2016-08-08", rating: 3})
+      trip3 = RideShare::Trip.new({id: 10, start_time: start_time, end_time: nil, driver: driver, cost: nil, passenger: nil, date: "2016-08-010", rating: nil})
+      driver.add_trip(trip1)
```
Comment (droberts-sea) [inline specs/driver_spec.rb:106]: A lot of this code is duplicated between tests. It might make sense to move it do a `begin` block up in the highest level `describe`, which would make it available in all sub-`describe`s.

File: specs/passenger_spec.rb [hunk 5]
```diff
@@ -74,5 +74,32 @@
         driver.must_be_kind_of RideShare::Driver
       end
     end
+  end#end get_drivers method
+
+  describe 'total_trips_cost method' do
+    it 'can return total money passenger spent on rides' do
+      @passenger = RideShare::Passenger.new(id: 9, name: "Merl Glover III", phone: "1-602-620-2330 x3723", trips: [])
```
Comment (droberts-sea) [inline specs/passenger_spec.rb:81]: The same types of tests are missing from `Passenger` as from `Driver`.

File: specs/trip_dispatch_spec.rb [hunk 6]
```diff
@@ -89,4 +90,44 @@
...
+      driver = @dispatcher.find_available_driver
+      driver.must_be_instance_of(RideShare::Driver)
+      driver.status.must_equal(:AVAILABLE)
+      driver.name.must_equal("Emory Rosenbaum")
+    end
+  end
+
+  describe 'request_trip method' do
+    before do
+      @dispatcher = RideShare::TripDispatcher.new
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:106]: What if there is no available driver? This is an important case, and is exactly the sort of situation where software will do something unexpected. You might set up the test as follows:

```ruby
it 'raises an error if no driver is available' do
  @dispatcher.drivers.each do |driver|
    # This may require an attr_writer in Driver
    driver.status = :UNAVAILABLE
  end
  proc {
    @dispatcher.request_trip(2)
  }.must_raise StandardError
end
```

**Review by droberts-sea — COMMENTED** [review]
FYI, I hit "submit" a little early on the feedback above, so it may look different now from what's in your email.
